src/Columns/MaskOperations.cpp

```diff
@@ -85,7 +85,11 @@ size_t extractMaskNumericImpl(
 {
     size_t ones_count = 0;
     size_t data_index = 0;
-    for (size_t i = 0; i != mask.size(); ++i)
+
+    size_t mask_size = mask.size();
+    size_t data_size = data.size();
+
+    for (size_t i = 0; i != mask_size && data_index != data_size; ++i)
```

Comment:
@Avogar below there is condition
```c++
if constexpr (column_is_short)
{
    index = data_index;
    ++data_index;
}
    else
         index = i;
```
I am not familiar with this part of short circuit evaluation, should not we always increate `data_index`, because we always increase index like:
```c++
if (!mask[i])
{
    ++data_index;
    continue;
}
```

Reply:
No, in case when we have `column_is_short=true` we expect that `data.size()` is equal to the number of ones in mask. And seems like this expectation could be wrong and it leads to crash. It means that the real bug is not here. Let's add your fix and add new check after the loop that `i == mask_size` and if it's not, throw logical error. I will try to find the real bug

Reply:
Yes, bug already fixed in https://github.com/ClickHouse/ClickHouse/pull/35500, because it returned wrong constant this parts rely on this. No changes required.